src/excmds.ts

```diff
@@ -4389,289 +4587,216 @@ export async function hint(option?: string, selectors?: string, ...rest: string[
-                    // https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes
-                    anchorUrl.hash = link.id || link.name
-                    // /!\ Warning: This is racy! This can easily be fixed by adding an await but do we want this? yank can be pretty slow, especially with yankto=selection
-                    run_exstr("yank " + anchorUrl.href)
-                    return link
-                },
-                rapid,
-            )
-            break
+            case OpenMode.YankAnchor:
+                hintables = hinting.toHintablesArray(DOM.anchors(includeInvisible))
+                break
 
-        case "-W":
-            selectHints = hinting.pipe(
-                DOM.HINTTAGS_selectors,
-                elem => {
-                    // /!\ RACY RACY RACY!
-                    run_exstr(selectors + " " + rest.join(" ") + " " + elem)
-                    return elem
-                },
-                rapid,
-                jshints,
-            )
-            break
+            case OpenMode.Images:
+            case OpenMode.ImagesTab:
+            case OpenMode.SaveImage:
+            case OpenMode.SaveAsImage:
+                // TODO: Support custom image selectors?
+                hintables = hinting.toHintablesArray(hinting.hintableImages(includeInvisible))
+                break
 
-        case "-pipe":
-            selectHints = hinting.pipe(selectors, elem => elem[rest.join(" ")], rapid, jshints)
-            break
+            case OpenMode.Kill:
+            case OpenMode.KillTridactyl:
+                // TODO: Support custom killable selectors?
+                hintables = hinting.toHintablesArray(hinting.killables(includeInvisible))
+                break
 
-        case "-i":
-            selectHints = hinting.pipe_elements(
-                hinting.hintableImages(),
-                elem => {
-                    open(new URL(elem.getAttribute("src"), window.location.href).href)
-                    return elem
-                },
-                rapid,
-            )
-            break
+            case OpenMode.SaveResource:
+            case OpenMode.SaveAsResource:
+                // TODO: Support custom saveable selectors?
+                hintables = hinting.toHintablesArray(hinting.saveableElements(includeInvisible))
+                break
 
-        case "-I":
-            selectHints = hinting.pipe_elements(
-                hinting.hintableImages(),
-                async elem => {
-                    await hintTabOpen(new URL(elem.getAttribute("src"), window.location.href).href)
-                    return elem
-                },
-                rapid,
-            )
-            break
+            default:
+                hintables = hinting.hintables(positionals.length ? positionals.join(" ") : DOM.HINTTAGS_selectors, jshints, includeInvisible)
```

Comment:
If selectors exist they should always override whatever the defaults are.

But how do we know that the positionals aren't e.g. callbacks?

Reply:
Ok, so this was not the case with the current implementation, at least for the "non-default" modes, like ones that operate on killables, images, and saveables. It is relatively easy to add though.

From the argument parsing, `positionals` are the "left over" arguments which were not associated with a flag (e.g. `-f`, `-fr`, `-W`, or `-F`). Each flag that takes a value changes the current state to one that will process the remaining flags accordingly (specifically, `-F` and `-W` will consume all the remaining arguments and join them to form a callback, as the previous implementation was doing), which leaves only selectors in the `positionals` array.

So `bind a -F some js` will result in `positionals = ["a"]` and `callback = "some js"`, and trigger the JS callback with the `a` selector.

Reply:
`hint a -F some js` instead of `hint -cF [selector] [callback]` will break a lot of user's code.

Including some of our own: https://github.com/tridactyl/tridactyl/blob/49c887222b72f7848957ba173a80037c49f29f57/src/lib/config.ts#L288

How would this handle `:hint a -F e => tri.excmds.tabopen("-b", e.href)`. Are we sure it wouldn't try to steal the `-b`?

---

I'm beginning to think that it's not worth trying to preserve backwards compatibility and instead call this function `hint2`. We can call the current one `hint_deprecated`, alias it to `hint` and use a config updater to try to change everyone's hint binds to `hint_deprecated`. In a few months we can do a similar dance and make `hint2` point to `hint`.

Reply:
I based the initial implementation on what I could make out from the documentation and how "usual" argument parsing happens in CLIs, where in combined short-form flags, if one flag takes a value, that value has to be the next argument passed to the command. I was specifically looking for feedback on this by opening this PR, since I'm mostly using Tridactyl without bindings so I'm curious what users came up with to have some test cases.

I can change the parser to handle the `-cF` case (two short arguments taking values), it just requires an extra state and this will provide compatibility. I think we're pretty close to full backwards compatibility (especially after implementing this change), but I'll leave the choice of whether or not to feature-gate it up to you!

For the example you describe, `:hint a -F e => tri.excmds.tabopen("-b", e.href)` will go through the following parser steps (parser state in brackets):
* `[Initial] ['a', ...]`: lone argument, add to positionals (thus `a` is a selector part)
* `[Initial] [..., '-F', ...]`: single argument `-F`, transition to `ExpectCallback`
* `[ExpectCallback] [..., 'e', '=>', 'tri.excmds.tabopen("-b",', 'e.href)']`: in ExpectCallback, consume all the remaining arguments into a callback unconditionally and exit the parsing loop

Thus, `positionals = ['a']` and `callback = 'e => tri.excmds.tabopen("-b", e.href)'`.

Reply:
Thanks for the link, I added some interesting examples I could find to the automated tests of the argument parser I added in https://github.com/tridactyl/tridactyl/pull/3534/commits/806760ca1c8f06b2cb1255c176dc0763a10c6631